grails-app/services/org/pih/warehouse/requisition/RequisitionService.groovy

```diff
@@ -859,4 +860,31 @@ class RequisitionService {
                 break
         }
     }
+
+    void deleteEvent(Requisition requisition, Event event) {
+        requisition.removeFromEvents(event)
+        event.delete()
+        requisition.save()
+    }
+
+    void rollbackLastEvent(Requisition requisition) {
+        def g = grailsApplication.mainContext.getBean('org.codehaus.groovy.grails.plugins.web.taglib.ApplicationTagLib')
+        Event event = requisition.mostRecentEvent
+        if (!event) {
+            throw new RuntimeException(g.message(
+                    code: "requisition.error.rollback.noRecentEvents",
+                    default: "Cannot rollback requisition because there are no recent events"
+            ))
+        }
+        try {
+            deleteEvent(requisition, event)
+        } catch (Exception e) {
```

Comment:
I think this catch won't do anything, since the eventual commit of the transaction (and the error that might happen then) from `deleteEvent` would happen at the end of the session?

Reply:
The main purpose of this catch is to show a more meaningful error, with: "Error rolling back most recent event".

Reply:
I know what's the purpose of that, but I meant that you won't catch any exception here, because any error that would occur from `deleteEvent` would happen at the end of the session - take a look at it is done in the same transaction - it is not the same case if you had `try/catch` in the controller and persistence actions in the service - here, you have try/catch in the service.

Reply:
ok, now it's clear